examples/scripts/lsrtm_example_supershot.jl

```diff
@@ -0,0 +1,226 @@
+
+# Generate data via nonlinear modeling
+dobs = Pr*F*adjoint(Ps)*q
+dobs0 = Pr*F0*adjoint(Ps)*q
+
+# Get linearize data for LS-RTM
+dlin = dobs-dobs0
+
+# Preconditioner
+idx_wb = find_water_bottom(m-m0)
+Tm = judiTopmute(model0.n, idx_wb, 3)  # Mute water column
+S = judiDepthScaling(model0)
+Mr = Tm*S
+
+# Linearized Bregman parameters
+
+batchsize = 2       # num of sources in each iteration
+niter = 4           # num of iterations (2 data pass)
+
+fval = zeros(Float32, niter)
+
+# Soft thresholding functions
+soft_thresholding(x::Array{Float64}, lambda) = sign.(x) .* max.(abs.(x) .- convert(Float64, lambda), 0.0)
+soft_thresholding(x::Array{Float32}, lambda) = sign.(x) .* max.(abs.(x) .- convert(Float32, lambda), 0f0)
+
+# sparsifying transform (wavelet transform is used in this case)
+C = joDWT(n[1], n[2]; DDT=Float32)
+
+
+# set up linearized bregman variables
+x = zeros(Float32, size(C,2))   # primal
+z = zeros(Float32, size(C,1))   # dual
+t = zeros(Float32, niter)       # step length
+
+anti_chatter = true
+
+# set up anti-chattering variables, please check the SIAM 2020 paper for details
+if anti_chatter
+    flag_ = BitArray(undef,size(C,1))
+    sumsign = zeros(Float32,size(C,1))
+end
+
+# set up computational simultaneous source experiments
+# set up source geometry
+xsrc_sim = [[q.geometry.xloc[i][1] for i = 1:nsrc] for j = 1:batchsize]
+ysrc_sim = [[0.0f0] for j = 1:batchsize]
+zsrc_sim = [[q.geometry.zloc[i][1] for i = 1:nsrc] for j = 1:batchsize]
+sim_src_geometry = Geometry(xsrc_sim, ysrc_sim, zsrc_sim; dt=q.geometry.dt[1], t=q.geometry.t[1])
+ntComp_sim = get_computational_nt(sim_src_geometry, recGeometry, model)
+info_sim = Info(prod(n), batchsize, ntComp_sim)
+Ps_sim = judiProjection(info_sim, sim_src_geometry)
+
+# set up receiver geometry
+sim_rec_geometry = Geometry(dlin.geometry.xloc[1],dlin.geometry.yloc[1],dlin.geometry.zloc[1]; dt=dobs.geometry.dt[1], t=dlin.geometry.t[1], nsrc=batchsize)
+Pr_sim = judiProjection(info_sim, sim_rec_geometry)
+
+# set up modeling
+F0_sim = judiModeling(info_sim, model0; options=opt)
+
+
+# Main loop
+for j = 1:niter
+    # Set up random weights for current iteration
+    weights = randn(Float32,batchsize,nsrc)
+
+    # Create wavelet according to the random weights
+    wavelet_sim = Vector{Matrix{Float32}}(undef, batchsize)
+    for k = 1:batchsize
+        wavelet_sim[k] = zeros(Float32, q.geometry.nt[1], nsrc)
+        wavelet_sim[k][:,1:nsrc] = q.data[1]*weights[k:k,:]
+    end
+        
+    # super shot for current iteration
+    q_j = judiVector(sim_src_geometry, wavelet_sim)
+
+    # set up linearized modeling operator
+    J_j = judiJacobian(Pr_sim*F0_sim*adjoint(Ps_sim), q_j)
+
+    # superposition of shot records
+    data_j = Vector{Matrix{Float32}}(undef, batchsize)
+    for k = 1:batchsize
+        data_j[k] = sum(weights[k,:].*dlin.data)
+    end
+
+    # simultaneous shot record for current iteration
+    dlin_j = judiVector(sim_rec_geometry, data_j)
```

Comment:
That's a lot of array re-creation and stuff, a god utility function (or even better a linear operator) would be much nicer. I.e `W*R*d` with `W` weights and `R` resitriction